diff --git a/infra/kernel_compat.h b/infra/kernel_compat.h
index 5b10ee7..c7e4714 100644
--- a/infra/kernel_compat.h
+++ b/infra/kernel_compat.h
@@ -1,6 +1,23 @@
 #ifndef __KSU_H_KERNEL_COMPAT
 #define __KSU_H_KERNEL_COMPAT
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 4, 0)
+
+#ifndef __has_attribute
+#define __has_attribute(x) __GCC4_has_attribute_##x
+#define __GCC4_has_attribute___fallthrough__ 0
+#endif
+
+#ifndef fallthrough
+#if __has_attribute(__fallthrough__)
+#define fallthrough __attribute__((__fallthrough__))
+#else
+#define fallthrough do { } while (0) /* fallthrough */
+#endif
+#endif
+
+#endif // LINUX_VERSION_CODE < 5.4.0
+
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 9, 0)) && (LINUX_VERSION_CODE < KERNEL_VERSION(4, 10, 0)) ||              \
     (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 14, 0)) && (LINUX_VERSION_CODE < KERNEL_VERSION(4, 15, 0))
 #ifdef HISI_SELINUX_EBITMAP_RO
@@ -11,6 +28,10 @@
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4, 10, 0) || defined(CONFIG_IS_HW_HISI) ||                                     \
     defined(CONFIG_KSU_ALLOWLIST_WORKAROUND)
 
+#ifndef strncpy_from_user_nofault
+#define strncpy_from_user_nofault ksu_strncpy_from_user_nofault
+#endif
+
 extern int install_session_keyring_to_cred(struct cred *cred, struct key *keyring);
 static struct key *init_session_keyring = NULL;
 
diff --git a/runtime/ksud.c b/runtime/ksud.c
index c9d11da..ad18fe5 100644
--- a/runtime/ksud.c
+++ b/runtime/ksud.c
@@ -61,7 +61,10 @@ void on_post_fs_data(void)
 #else
     stop_input_hook();
 #endif
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
     ksu_selinux_hide_handle_post_fs_data();
+#endif
 }
 
 extern void ext4_unregister_sysfs(struct super_block *sb);
@@ -223,7 +226,9 @@ int ksu_handle_execveat_ksud(int *fd, struct filename **filename_ptr,
             char buf[16];
             if (check_argv(*argv, 1, "second_stage", buf, sizeof(buf))) {
                 pr_info("/system/bin/init second_stage executed\n");
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 0)
                 ksu_selinux_hide_handle_second_stage();
+#endif
                 apply_kernelsu_rules();
                 cache_sid();
                 setup_ksu_cred();
diff --git a/selinux/selinux.c b/selinux/selinux.c
index 1f8fd0e..4183b1c 100644
--- a/selinux/selinux.c
+++ b/selinux/selinux.c
@@ -317,4 +317,4 @@ void susfs_set_batch_sid(void)
     susfs_set_sid(KERNEL_INIT_DOMAIN, &susfs_init_sid);
     susfs_set_sid(KERNEL_PRIV_APP_DOMAIN, &susfs_priv_app_sid);
 }
-#endif
\ No newline at end of file
+#endif
